File: apps/web/src/app/(private)/_lib/createSdk.ts
import { NotFoundError } from '@latitude-data/constants/errors'
import { compactObject } from '@latitude-data/core/lib/compactObject'
import { Result } from '@latitude-data/core/lib/Result'
import { findFirstApiKey } from '@latitude-data/core/queries/apiKeys/findFirst'
import { env } from '@latitude-data/env'
import { Latitude } from '@latitude-data/sdk'
import { LogSources } from '@latitude-data/core/constants'

import { Workspace } from '@latitude-data/core/schema/models/types/Workspace'
// NOTE: this would be a great candidate for a cache function with redis
async function getLatitudeApiKey(workspace: Workspace) {
  const firstApiKey = await findFirstApiKey({ workspaceId: workspace.id })

  if (!firstApiKey) {
    return Result.error(
      new NotFoundError("Couldn't find a valid Latitude API key"),
    )
  }

  return Result.ok(firstApiKey)
}

export async function createSdk({
  workspace,
  projectId,
  apiKey,
  __internal,
}: {
  workspace: Workspace
  projectId?: number
  apiKey?: string
  __internal?: { source: LogSources }
}) {
  if (!apiKey) {
    const result = await getLatitudeApiKey(workspace)
    if (result.error) return result

    apiKey = result.value.token
  }

  const gateway = {
    host: env.GATEWAY_HOSTNAME,
    port: env.GATEWAY_PORT,
    ssl: env.GATEWAY_SSL,
  }

  return Result.ok(
    new Latitude(
      apiKey,
      compactObject({
        gateway,
        projectId,
        __internal: {
          ...__internal,
          gateway,
        },
      }),
    ),
  )
}

File: examples/sdks/python/util/create_sdk.py
from typing import Optional
from dotenv import load_dotenv
import os
from latitude_sdk import Latitude, LatitudeOptions, InternalOptions, GatewayOptions

load_dotenv()

LATITUDE_API_KEY = os.getenv("LATITUDE_API_KEY")
GATEWAY_HOST = os.getenv("GATEWAY_HOST")
GATEWAY_PORT = os.getenv("GATEWAY_PORT")

def create_sdk(options: Optional[LatitudeOptions] = None):
    assert LATITUDE_API_KEY, "LATITUDE_API_KEY is not set"
    gateway = None

    if (GATEWAY_HOST and GATEWAY_PORT):
        gateway = GatewayOptions(
            host=GATEWAY_HOST,
            port=int(GATEWAY_PORT),
            ssl=False,
            api_version="v3"
        )
    sdk = Latitude(
        LATITUDE_API_KEY,
        LatitudeOptions(
            project_id=options.project_id if options else None,
            version_uuid=options.version_uuid if options else None,
            internal=InternalOptions(gateway=gateway)
        )
    )
    return sdk
